e2e/cp_test.go

```diff
@@ -2926,3 +2926,311 @@ func TestCopyS3ToDirDryRun(t *testing.T) {
+		dst              string
+		wantedFile       string
+		expectedFiles    []string
+		nonExpectedFiles []string
+		rawFlag          string
+	}{
+		{
+			name: "cp --raw file*.txt s3://bucket/",
+			src: []fs.PathOp{
+				fs.WithFile("file*.txt", "content"),
+				fs.WithFile("file*1.txt", "content"),
+				fs.WithFile("file*file.txt", "content"),
+				fs.WithFile("file*2.txt", "content"),
+			},
+			wantedFile:       "file*.txt",
+			dst:              "s3://bucket/",
+			expectedFiles:    []string{"file*.txt"},
+			nonExpectedFiles: []string{"file*1.txt", "file*file.txt", "file*2.txt"},
+			rawFlag:          "--raw",
+		},
+		{
+			name: "cp  file*.txt s3://bucket/",
+			src: []fs.PathOp{
+				fs.WithFile("file*.txt", "content"),
+				fs.WithFile("file*1.txt", "content"),
+				fs.WithFile("file*file.txt", "content"),
+				fs.WithFile("file*2.txt", "content"),
+			},
+			wantedFile:       "file*.txt",
+			dst:              "s3://bucket/",
+			expectedFiles:    []string{"file*.txt", "file*1.txt", "file*file.txt", "file*2.txt"},
+			nonExpectedFiles: []string{},
+			rawFlag:          "",
+		},
+		{
+			name: "cp  a*/file.txt s3://bucket/",
+			src: []fs.PathOp{
+				fs.WithDir(
+					"a*",
+					fs.WithFile("file.txt", "content"),
+					fs.WithFile("file1.txt", "content"),
+				),
+				fs.WithDir(
+					"a*b",
+					fs.WithFile("file2.txt", "content"),
+					fs.WithFile("file3.txt", "content"),
+				),
+
+				fs.WithFile("file4.txt", "content"),
+			},
+			wantedFile:       "a*/file.txt",
+			dst:              "s3://bucket/",
+			expectedFiles:    []string{"file.txt"}, // when full path entered, the base part is uploaded.
+			nonExpectedFiles: []string{"a*/file.txt", "a*/file1.txt", "a*b/file2.txt", "a*/file3.txt", "file4.txt", "file1.txt", "file2.txt", "file3.txt"},
+			rawFlag:          "--raw",
+		},
+	}
+
+	for _, tc := range testcases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			s3client, s5cmd, cleanup := setup(t)
+			defer cleanup()
+
+			createBucket(t, s3client, bucket)
+
+			workdir := fs.NewDir(t, tc.name, tc.src...)
+			defer workdir.Remove()
+
+			srcpath := filepath.ToSlash(workdir.Join(tc.wantedFile))
+
+			cmd := s5cmd("cp", srcpath, tc.dst)
+			if tc.rawFlag != "" {
+				cmd = s5cmd("cp", tc.rawFlag, srcpath, tc.dst)
+			}
+
+			result := icmd.RunCmd(cmd)
+			result.Assert(t, icmd.Success)
+
+			// check files in S3
+			for _, obj := range tc.expectedFiles {
+				err := ensureS3Object(s3client, bucket, obj, "content")
+				if err != nil {
+					t.Fatalf("%s is not exist in s3\n", obj)
+				}
+			}
+
+			// check files not in S3
+			for _, obj := range tc.nonExpectedFiles {
+				err := ensureS3Object(s3client, bucket, obj, "content")
+				assertError(t, err, errS3NoSuchKey)
+			}
+
+			// assert filesystem
+			expected := fs.Expected(t, tc.src...)
+			assert.Assert(t, fs.Equal(workdir.Path(), expected))
+		})
+	}
+
+}
+
+// When folder is uploaded with --raw flag, it gives an error.
```

Comment:
Why? I think it should copy files in `a*` directory in raw mode.

Reply:
Thanks for the feedback. I changed as you suggested and here is my thoughts.

1. Now we can have wildcard in the destination in `--raw` mode. 

2. Changing URL package is simpler than adding new functions or changing `expandSource` function. It is working good and adding `--raw` flag to other commands is much more easier now. 

3. Added tests for remote -> remote. 

Also, local directory `a*` is uploaded now without uploading other folders matching `a*` pattern.